execution-engine/src/wasi/wasmi.rs

```diff
@@ -0,0 +1,1203 @@
+            WASIAPIName::FD_FILESTAT_SET_SIZE => self.wasi_fd_filestat_set_size(args),
+            WASIAPIName::FD_FILESTAT_SET_TIMES => self.wasi_fd_filestat_set_times(args),
+            WASIAPIName::FD_PREAD => self.wasi_fd_pread(args),
+            WASIAPIName::FD_PRESTAT_GET => self.wasi_fd_prestat_get(args),
+            WASIAPIName::FD_PRESTAT_DIR_NAME => self.wasi_fd_prestat_dir_name(args),
+            WASIAPIName::FD_PWRITE => self.wasi_fd_pwrite(args),
+            WASIAPIName::FD_READ => self.wasi_fd_read(args),
+            WASIAPIName::FD_READDIR => self.wasi_fd_readdir(args),
+            WASIAPIName::FD_RENUMBER => self.wasi_fd_renumber(args),
+            WASIAPIName::FD_SEEK => self.wasi_fd_seek(args),
+            WASIAPIName::FD_SYNC => self.wasi_fd_sync(args),
+            WASIAPIName::FD_TELL => self.wasi_fd_tell(args),
+            WASIAPIName::FD_WRITE => self.wasi_fd_write(args),
+            WASIAPIName::PATH_CREATE_DIRECTORY => self.wasi_path_create_directory(args),
+            WASIAPIName::PATH_FILESTAT_GET => self.wasi_path_filestat_get(args),
+            WASIAPIName::PATH_FILESTAT_SET_TIMES => self.wasi_path_filestat_set_times(args),
+            WASIAPIName::PATH_LINK => self.wasi_path_link(args),
+            WASIAPIName::PATH_OPEN => self.wasi_path_open(args),
+            WASIAPIName::PATH_READLINK => self.wasi_path_readlink(args),
+            WASIAPIName::PATH_REMOVE_DIRECTORY => self.wasi_path_remove_directory(args),
+            WASIAPIName::PATH_RENAME => self.wasi_path_rename(args),
+            WASIAPIName::PATH_SYMLINK => self.wasi_path_symlink(args),
+            WASIAPIName::PATH_UNLINK_FILE => self.wasi_path_unlink_file(args),
+            WASIAPIName::POLL_ONEOFF => self.wasi_poll_oneoff(args),
+            WASIAPIName::PROC_EXIT => self.wasi_proc_exit(args),
+            WASIAPIName::PROC_RAISE => self.wasi_proc_raise(args),
+            WASIAPIName::SCHED_YIELD => self.wasi_sched_yield(args),
+            WASIAPIName::RANDOM_GET => self.wasi_random_get(args),
+            WASIAPIName::SOCK_RECV => self.wasi_sock_recv(args),
+            WASIAPIName::SOCK_SEND => self.wasi_sock_send(args),
+            WASIAPIName::SOCK_SHUTDOWN => self.wasi_sock_shutdown(args),
+        }?;
+        Ok(Some(RuntimeValue::I32((return_code as i16).into())))
+    }
+}
+
+/// Functionality of the `WASMIRuntimeState` type that relies on it satisfying
+/// the `Externals` and `ModuleImportResolver` constraints.
+impl WASMIRuntimeState {
+    /// Creates a new initial `HostProvisioningState`.
+    pub fn new(filesystem: Arc<Mutex<FileSystem>>, program_name: &str) -> Self {
+        Self {
+            vfs: WASIWrapper::new(filesystem, Principal::Program(program_name.to_string())),
+            program: Principal::NoCap,
+            program_module: None,
+            memory: None,
+        }
+        //NOTE: cannot find a way to immediately call the load_program here,
```

Comment:
Is this something that should be added to the issue tracker?

Reply:
I think it is more about simplicity of code. It was a bit annoying about constantly dereferencing the memory. I will not consider an issue we should track.